plugin/src/main/scala/org/virtuslab/akkasaferserializer/SaferSerializerPluginComponent.scala

```diff
@@ -0,0 +1,64 @@
+package org.virtuslab.akkasaferserializer
+
+import scala.tools.nsc.plugins.PluginComponent
+import scala.tools.nsc.{Global, Phase}
+
+class SaferSerializerPluginComponent(val pluginOptions: PluginOptions, val global: Global) extends PluginComponent {
+  import global._
+  override val phaseName: String = "akka-safer-serializer-gather"
+  override val runsAfter: List[String] = List("refchecks")
+
+  var annotatedTraitsCache: List[Type] = List()
+
+  override def newPhase(prev: Phase): Phase =
+    new StdPhase(prev) {
+      override def apply(unit: global.CompilationUnit): Unit = {
+        val body = unit.body
+        val reporter = CrossVersionReporter(global)
+
+        val genericsNames = Seq(
+          "akka.actor.typed.Behavior",
+          "akka.persistence.typed.scaladsl.ReplyEffect",
+          "akka.projection.eventsourced.EventEnvelope")
+
+        annotatedTraitsCache = body
+          .collect {
+            case x: TypeTree if genericsNames.contains(x.tpe.typeSymbol.fullName) => x.tpe.typeArgs
+          }
+          .flatten
+          .foldRight(annotatedTraitsCache) { (next, annotatedTraits) =>
+            if (annotatedTraits.exists(next <:< _) || next.typeSymbol.fullName.startsWith("akka")) {
+              annotatedTraits
+            } else {
+              findSuperclassAnnotatedWithSerializabilityTrait(next) match {
+                case Some(annotatedType) =>
+                  if (pluginOptions.verbose)
+                    reporter.echo(
+                      s"${classOf[SaferSerializerPlugin].getSimpleName}: Found new annotated trait: ${annotatedType.typeSymbol.fullName}")
+                  annotatedType :: annotatedTraits
+                case None =>
+                  reporter.error(
```

Comment:
Hmm still getting false positives here (almost exclusively on upper bounds of type params)... details on Slack since they contain pieces of proprietary code